### PR TITLE
CB improvements

### PR DESCRIPTION
**Pull request by ilya-lavrenov**

### CODE REVIEW

File: src/cpp/src/block_manager.hpp [hunk 1]
```diff
@@ -246,42 +247,43 @@ class BlockManager {
...
-        auto running_sequences = sequence_group->get_not_finished_sequences();
-        std::set<size_t> blocks_released_indices;
+    // frees at most 'num_required_blocks' blocks for a given sequence group and
+    // returns actually de-allocated number of KV cache blocks
+    // This method is used by preemption algorigthm
+    const size_t free_sequence_group_partially(SequenceGroup::Ptr sequence_group, size_t num_required_blocks) {
+        auto running_sequences = sequence_group->get_running_sequences();
+        // TODO: ilavrenov - why beam search case is not handled here?
+        // in case of beam search 'blocks_num' are not equally distributed between sequences
+        // because some of them share the same blocks
```
Comment (ilya-lavrenov) [inline src/cpp/src/block_manager.hpp:263]: CC @popovaan

File: src/cpp/src/scheduler.hpp [hunk 2]
```diff
@@ -86,63 +89,63 @@ class Scheduler {
...
         }
         preempted_tokens = tokens_in_last_block + std::max<size_t>((int)logical_blocks_released - 1, 0) * block_size;
 
         // case when preemption requires preempt prompt tokens
-        if (!m_config.dynamic_split_fuse && processed_tokens - preempted_tokens < sequence_group->get_prompt_len()) {
+        if (!m_config.dynamic_split_fuse && context_len - preempted_tokens < sequence_group->get_prompt_len()) {
             // preempt prompt fully to not leave partially generated prompt
-            preempted_tokens = processed_tokens;
+            preempted_tokens = context_len;
+            // TODO: ilavrenov - what is we have multiple sequences within a group? we need to drop them all
```
Comment (ilya-lavrenov) [inline src/cpp/src/scheduler.hpp:135]: CC @popovaan

File: src/cpp/include/openvino/genai/generation_handle.hpp [hunk 3]
```diff
@@ -52,37 +53,53 @@ struct GenerationResult {
     GenerationStatus m_status = GenerationStatus::RUNNING;
 };
 
+// Represents already generated tokens of running generate() method.
+// E.g. typically generate() method consists of multiple step() which generate
+// token by token. This structure represents a vector of already generated tokens so far
+// for a given prompt.
```
Comment (mzegla) [inline src/cpp/include/openvino/genai/generation_handle.hpp:59]: That's not exactly true. I would rephrase that to: "... a vector of tokens generated since the last read...". 
To be more specific it's always vector of one element if N == 1 and we are not using beam search. 
For N > 1 and/or beam search is used vector contains all generated tokens.

File: src/cpp/include/openvino/genai/generation_handle.hpp [hunk 4]
```diff
@@ -52,37 +53,53 @@ struct GenerationResult {
...
     GenerationFinishReason finish_reason;
 };
 
+// Current outputs of step() method for all scheduled requests
 using GenerationOutputs = std::unordered_map<uint64_t, GenerationOutput>;
 
 class GenerationStream;
 
-class OPENVINO_GENAI_EXPORTS GenerationHandleImpl {
+class OPENVINO_GENAI_EXPORTS GenerationHandle {
```
Comment (mzegla) [inline src/cpp/include/openvino/genai/generation_handle.hpp:76]: Why do we need this change?

File: src/cpp/include/openvino/genai/generation_handle.hpp [hunk 5]
```diff
@@ -52,37 +53,53 @@ struct GenerationResult {
...
     std::shared_ptr<GenerationStream> m_generation_stream;
     ov::genai::GenerationConfig m_sampling_params;
 
-    bool is_dropped();
+    // whether client ha dropped session with pipeline
```
Comment (mzegla) [inline src/cpp/include/openvino/genai/generation_handle.hpp:80]:
```suggestion
    // whether client has dropped session with pipeline
```

File: src/cpp/include/openvino/genai/generation_handle.hpp [hunk 6]
```diff
@@ -52,37 +53,53 @@ struct GenerationResult {
...
-    GenerationHandleImpl(const GenerationHandleImpl&) = delete;
-    GenerationHandleImpl& operator=(const GenerationHandleImpl&) = delete;
+    GenerationHandle(const GenerationHandle&) = delete;
+    GenerationHandle& operator=(const GenerationHandle&) = delete;
 
-    GenerationStatus get_status();
+    GenerationStatus get_status() const;
 
-    bool can_read();
+    // whether new tokens are available
```
Comment (mzegla) [inline src/cpp/include/openvino/genai/generation_handle.hpp:99]:
```suggestion
    // whether read() is possible (new tokens are available and handle has not been dropped)
```

File: src/cpp/include/openvino/genai/scheduler_config.hpp [hunk 7]
```diff
@@ -9,34 +9,40 @@ namespace ov::genai {
...
 
     // block size for KV cache
     std::size_t block_size = 32;
 
     // whether to split prompt / generate to different scheduling phases
+    // - Dynamic split fuse schdules requests in generation phase first, then
+    // schdules requests in prompt phase. If request cannot be fully fit into
+    // remaining space of 'max_num_batched_tokens' group, it's scheduled only partially
+    // and other tokens can be scheduled only next iterations
+    // - vLLM mode priorities requests in prompt phase over requests on generation phase
```
Comment (mzegla) [inline src/cpp/include/openvino/genai/scheduler_config.hpp:30]:
```suggestion
    // whether to split prompt / generate to different scheduling phases
    // - Dynamic split fuse schedules requests in generation phase first, then
    // schedules requests in prompt phase. If request cannot be fully fit into
    // remaining space of 'max_num_batched_tokens' group, it's scheduled only partially
    // and other tokens can be scheduled in next iterations
    // - vLLM mode prioritizes requests in prompt phase over requests in generation phase
```
